tools/flow_grid.py

```diff
@@ -0,0 +1,689 @@
+                data = self.dem
+
+        # generate grid of indices
+        indices = np.indices(self.shape, dtype=self.shape_min)
+
+        # handle nodata values in dem
+        dem_nodata = self.nodata['dem']
+        dem_mask = (data == dem_nodata)        
+        np.place(data, dem_mask, np.iinfo(data.dtype.type).max)
+
+        # initialize indices of corners
+        corner = {
+        'nw' : {'k' : tuple(indices[:,0,0]),
+                'v' : [[0,1,1], [1,1,0]],
+                'pad': np.array([3,4,5])},
+        'ne' : {'k' : tuple(indices[:,0,-1]),
+                'v' : [[1,1,0], [-1,-2,-2]],
+                'pad': np.array([5,6,7])},
+        'sw' : {'k' : tuple(indices[:,-1,0]),
+                'v' : [[-2,-2,-1], [0,1,1]],
+                'pad': np.array([1,2,3])},
+        'se' : {'k' : tuple(indices[:,-1,-1]),
+                'v' : [[-1,-2,-2], [-2,-2,-1]],
+                'pad': np.array([7,8,1])}
+        }
+    
+        # initialize indices of edges
+        edge = {
+        'n' : {'k' : tuple(indices[:,0,1:-1]),
+               'pad' : np.array([3,4,5,6,7])},
+        'w' : {'k' : tuple(indices[:,1:-1,0]),
+               'pad' : np.array([1,2,3,4,5])},
+        'e' : {'k' : tuple(indices[:,1:-1,-1]),
+               'pad' : np.array([1,5,6,7,8])},
+        's' : {'k' : tuple(indices[:,-1,1:-1]),
+               'pad' : np.array([1,2,3,7,8])}
+        }
+    
+        # initialize indices of body (all cells except edges and corners)
+        body = indices[:, 1:-1, 1:-1]
+    
+        # initialize output array
+        outmap = np.full(self.shape, nodata, dtype=np.int8)
+    
+        # select the eight cells surrounding a cell
+        def select_surround(i, j):
+            return ([i-1, i-1, i+0, i+1, i+1, i+1, i+0, i-1],
+                   [j+0, j+1, j+1, j+1, j+0, j-1, j-1, j-1])
+    
+        # select the five cells surrounding an edge cell
+        def select_edge_sur(k):
+            i,j = edge[k]['k']
+            if k == 'n':
+                return [i+0, i+1, i+1, i+1, i+0], [j+1, j+1, j+0, j-1, j-1]
+            elif k =='e':
+                return [i-1, i+1, i+1, i+0, i-1], [j+0, j+0, j-1, j-1, j-1]
+            elif k =='s':
+                return [i-1, i-1, i+0, i+0, i-1], [j+0, j+1, j+1, j-1, j-1]
+            elif k == 'w':
+                return [i-1, i-1, i+0, i+1, i+1], [j+0, j+1, j+1, j+1, j+0]
+     
+        # for each entry in "body" determine flow direction based
+        # on steepest neighboring slope
+        for i, j in np.nditer(tuple(body), flags=['external_loop']):
+            dat = data[i,j]
+            sur = data[select_surround(i,j)]
+            a = ((dat - sur) > 0).any(axis=0)
+            b = np.argmax((dat - sur), axis=0) + 1
+            c = flat
+            outmap[i,j] = np.where(a,b,c)
+
+        # determine flow direction for edges and corners, if desired
+        if include_edges == True:
+
+            # fill corners
+            for i in corner.keys():
```

Comment:
this can just be `for i in corner`:

Also, don't use `i` and `j` for variables other than grid indicies.  

Maybe this would be best as `for corner in corners:` or something similar.